packages/@aws-cdk/aws-lambda/lib/adot-layers.ts

```diff
@@ -0,0 +1,240 @@
+
+  /**
+   * Version 1.19.0
+   */
+  public static readonly V1_19_0 = new AdotLambdaLayerJavaSdkVersion('1.19.0');
+}
+
+/**
+ * The collection of versions of the ADOT Lambda Layer for Java auto-instrumentation
+ */
+export class AdotLambdaLayerJavaAutoInstrumentationVersion extends AdotLambdaLayerVersionBase {
+  /**
+   * Will always use the latest layer version. New versions could
+   * introduce incompatible changes. Make sure to test new versions.
+   */
+  public static readonly LATEST = new AdotLambdaLayerJavaAutoInstrumentationVersion('1.19.2');
+
+  /**
+   * Version 1.19.2
+   */
+  public static readonly V1_19_2 = new AdotLambdaLayerJavaAutoInstrumentationVersion('1.19.2');
+}
+
+/**
+ * The collection of versions of the ADOT Lambda Layer for Python SDK
+ */
+export class AdotLambdaLayerPythonSdkVersion extends AdotLambdaLayerVersionBase {
+  /**
+   * Will always use the latest layer version. New versions could
+   * introduce incompatible changes. Make sure to test new versions.
+   */
+  public static readonly LATEST = new AdotLambdaLayerPythonSdkVersion('1.13.0');
+
+  /**
+   * Version 1.13.0
+   */
+  public static readonly V1_13_0 = new AdotLambdaLayerPythonSdkVersion('1.13.0');
+}
+
+/**
+ * The collection of versions of the ADOT Lambda Layer for JavaScript SDK
+ */
+export class AdotLambdaLayerJavaScriptSdkVersion extends AdotLambdaLayerVersionBase {
+  /**
+   * Will always use the latest layer version. New versions could
+   * introduce incompatible changes. Make sure to test new versions.
+   */
+  public static readonly LATEST = new AdotLambdaLayerJavaScriptSdkVersion('1.7.0');
+
+  /**
+   * Version 1.7.0
+   */
+  public static readonly V1_7_0 = new AdotLambdaLayerJavaScriptSdkVersion('1.7.0');
+}
+
+/**
+ * The collection of versions of the ADOT Lambda Layer for generic purpose
+ */
+export class AdotLambdaLayerGenericVersion extends AdotLambdaLayerVersionBase {
+  /**
+   * Will always use the latest layer version. New versions could
+   * introduce incompatible changes. Make sure to test new versions.
+   */
+  public static readonly LATEST = new AdotLambdaLayerGenericVersion('0.62.1');
+
+  /**
+   * Version 0.62.1
+   */
+  public static readonly V0_62_1 = new AdotLambdaLayerGenericVersion('0.62.1');
+}
+
+/**
+ * Properties for an ADOT instrumentation in Lambda
+ */
+export interface AdotInstrumentationProps {
+  /**
+   * The type of the ADOT Lambda layer.
+   */
+  readonly type: AdotLambdaLayerType;
+
+  /**
+   * The startup script to run, see ADOT documentation to pick the right script for your use case: https://aws-otel.github.io/docs/getting-started/lambda
+   */
+  readonly execWrapper: string;
+
+  /**
+   * The version of the Lambda layer
+   */
+  readonly version:
+  | AdotLambdaLayerJavaSdkVersion
+  | AdotLambdaLayerJavaAutoInstrumentationVersion
+  | AdotLambdaLayerJavaScriptSdkVersion
+  | AdotLambdaLayerPythonSdkVersion
+  | AdotLambdaLayerGenericVersion;
```

Comment:
jsii doesn't allow union types (I'm not sure why the build is passing actually).
What if we combined the type with the version?

```ts
export abstract class AdotLambdaLayerType {
  public static fromJavaSdkVersion(version: AdotLambdaLayerJavaSdkVersion, execWrapper: string): AdotLambdaLayerType {
    return new JavaSdkVersionLayer({
      version: version.version,
      type: AdotLayerType.JAVA_SDK,
    });
  }

  public abstract bind(scope: Construct, handler: IFunction, options?: LayerBindOptions): AdotLambdaLayerConfig;
}

class JavaSdkVersionLayer extends AdotLambdaLayerType {
  constructor(private readonly options: JavaSdkVersionLayerOptions) {
    super();
  }

  public bind(scope: Construct, handler: IFunction, options: LayerBindOptions = {}): AdotLambdaLayerConfig {
    // can do the individual runtime checks here, just an example
    if (handler.runtime === Runtime.JAVA_8) {
      throw new Error(...);
    }	
    return {
      version: this.options.version,
      architecture: handler.architecture,
      type: this.options.type,
    };
  }
}
```

Then in the addAdotLayer method you can just do
```ts
const config = props.type.bind(this, this);
AdotLambdaLayer.fromLayerAttributes(this, 'Layer', config);
```

The user would do something like

```ts
handler.addAdotLayer({
  type: AdotLambdaLayerType.fromJavaSdkVersion(AdotLambdaLayerJavaSdkVersion.V1_9_1),
})
```

Reply:
JSII does allow union types, it just does not work in other languages the same way as it does in TypeScript.

Since all those `*Version` classes inherits a base class, how about we just use that base class instead? My latest revision demonstrates the idea. But if you still prefer to have a specific function for each ADOT layer type (which I honestly want to avoid), I can do that.